File: day_4/puzzle_1/solve.py
import os

from benchmark import benchmark
import re


def main():
    def get_file_path():
        directory = os.path.dirname(os.path.abspath(__file__))
        return os.path.join(directory, "input.txt")

    file_path = get_file_path()

    with open(file_path, "r") as file:
        pairs = file.read().splitlines()

        count = 0

        for pair in pairs:
            start_a, end_a, start_b, end_b = map(int, re.split(r"[, -]", pair))

            if start_a <= start_b and end_a >= end_b:
                count += 1
            elif start_b <= start_a and end_b >= end_a:
                count += 1

        return count


if __name__ == "__main__":
    benchmark(main)

File: day_4/puzzle_1/solve.rb
# frozen_string_literal: true

require_relative '../../benchmark/benchmark'

Benchmark.run('day_4_puzzle_1') do
  File.open('./day_4/input.txt', 'r') do |file|
    pairs = file.read.split("\n")

    count = pairs.count do |pair|
      start_a, end_a, start_b, end_b = pair.split(/[, -]/).map(&:to_i)

      if start_a <= start_b && end_a >= end_b
        true
      elsif start_b <= start_a && end_b >= end_a
        true
      else
        false
      end
    end

    count
  end
end
